src/meta/expr.lean

```diff
@@ -580,6 +580,16 @@ meta def to_implicit_binder : expr → expr
 | (pi n _ d b) := pi n binder_info.implicit d b
 | e  := e
 
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def list_summands : expr → list expr
+| `(has_add.add %%a %%b) := a.list_summands ++ b.list_summands
+| a                      := [a]
+
+/--  Takes an `expr` and returns a list of its factors. -/
+meta def list_factors : expr → list expr
+| `(has_mul.mul %%a %%b) := a.list_summands ++ b.list_summands
+| a                      := [a]
```

Comment:
If we're not going to implement these in terms of `list_binary_operands` (let's see what @robertylewis thinks), then we should at least add a cross-reference to the docstrings.

Reply:
I'm okay with implementing these separately from `list_binary_operands` because they're not `tactic` valued. Morally `list_binary_operands` shouldn't be either, but the non-`tactic` implementation is a pain.